polkadot/xcm/xcm-builder/src/asset_exchange.rs

```diff
@@ -0,0 +1,178 @@
+);
+impl<AssetConversion, Fungibles, Matcher, AccountId> AssetExchange
+	for FungiblesPoolAdapter<AssetConversion, Fungibles, Matcher, AccountId>
+where
+	AssetConversion: SwapCredit<
+			AccountId,
+			Balance = u128,
+			AssetKind = Fungibles::AssetId,
+			Credit = fungibles::Credit<AccountId, Fungibles>,
+		> + QuoteExchangePrice<Balance = u128, AssetKind = Fungibles::AssetId>,
+	Fungibles: fungibles::Balanced<AccountId, Balance = u128>,
+	Matcher: MatchesFungibles<Fungibles::AssetId, Fungibles::Balance>,
+{
+	fn exchange_asset(
+		_: Option<&Location>,
+		give: AssetsInHolding,
+		want: &Assets,
+		maximal: bool,
+	) -> Result<AssetsInHolding, AssetsInHolding> {
+		let give_asset = give.fungible_assets_iter().next().ok_or_else(|| {
+			log::trace!(
+				target: "xcm::FungiblesPoolAdapter::exchange_asset",
+				"No fungible asset was in `give`.",
+			);
+			give.clone()
+		})?;
+		let want_asset = want.get(0).ok_or_else(|| {
+			log::trace!(
+				target: "xcm::FungiblesPoolAdapter::exchange_asset",
+				"No asset was in `want`.",
+			);
+			give.clone()
+		})?;
+		let (give_asset_id, give_amount) =
+			Matcher::matches_fungibles(&give_asset).map_err(|error| {
+				log::trace!(
+					target: "xcm::FungiblesPoolAdapter::exchange_asset",
+					"Could not map XCM asset {:?} to FRAME asset. Error: {:?}",
+					give,
+					error,
+				);
+				give.clone()
+			})?;
+		let (want_asset_id, want_amount) =
+			Matcher::matches_fungibles(&want_asset).map_err(|error| {
+				log::trace!(
+					target: "xcm::FungiblesPoolAdapter::exchange_asset",
+					"Could not map XCM asset {:?} to FRAME asset. Error: {:?}",
+					want,
+					error,
+				);
+				give.clone()
+			})?;
+
+		// We have to do this to convert the XCM assets into credit the pool can use.
+		let swap_asset = give_asset_id.clone().into();
+		let credit_in = Fungibles::issue(give_asset_id, give_amount);
+
+		// Do the swap.
+		let credit_out = if maximal {
+			// If `maximal`, then we swap exactly `credit_in` to get as much of `want_asset_id` as
+			// we can, with a minimum of `want_amount`.
+			<AssetConversion as SwapCredit<_>>::swap_exact_tokens_for_tokens(
+				vec![swap_asset, want_asset_id],
+				credit_in,
+				Some(want_amount),
+			)
+			.map_err(|(credit_in, _error)| {
+				// TODO: Log error.
+				drop(credit_in);
+				give.clone()
+			})?
+		} else {
+			// If `minimal`, then we swap as little of `credit_in` as we can to get exactly
+			// `want_amount` of `want_asset_id`.
+			let (credit_out, credit_change) =
+				<AssetConversion as SwapCredit<_>>::swap_tokens_for_exact_tokens(
+					vec![swap_asset, want_asset_id],
+					credit_in,
+					want_amount,
+				)
+				.map_err(|(credit_in, _)| {
+					drop(credit_in);
+					give.clone()
+				})?;
+
+			// TODO: If we want to make this a generic adapter, this need not be 0. Handle it.
+			// Probably depositing it back to the holding.
+			debug_assert!(credit_change.peek() == Zero::zero());
```

Comment:
I guess we can put the `credit_change`, and the `give` except first fungible asset, inside the result.

If performance is a downside for this, I can understand why you would implement only a very minimal version of the trait.

Reply:
Yeah given that we return `AssetsInHolding`, it makes sense to bundle them up together